src/main/java/org/takes/tk/AbstractTkProxy.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.takes.tk;
+
+import java.io.IOException;
+import lombok.EqualsAndHashCode;
+import org.takes.Request;
+import org.takes.Response;
+import org.takes.Take;
+
+/**
+ * Wrap of take.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Pierre Struve (pierre_struve@yahoo.fr)
+ * @version $Id$
+ * @since 1.0
+ */
+@EqualsAndHashCode(of = "origin")
+public abstract class AbstractTkProxy implements Take {
+
+    /**
+     * Original take.
+     */
+    private final transient Take origin;
+
+    /**
+     * Ctor.
+     * @param take Original take
+     */
+    public AbstractTkProxy(final Take take) {
```

Comment:
@tiaped instead of a `Take`, what we should accept in the ctor is the destination address and port. Something like `com.google.common.net.HostAndPort` but I'm not sure we can add that as a dependency. If not, we can do a manual host/port pair.